sparse-traits/src/spaces.rs

```diff
@@ -0,0 +1,78 @@
+//! Linear spaces and their elements.
+
+use crate::IndexSet;
+use crate::{Error, Result};
+use crate::{IndexType, Scalar};
+
+/// Definition of a linear space
+///
+/// Linear spaces are basic objects that can create
+/// elements of the space.
+pub trait LinearSpace {
+    /// Field Type.
+    type F: Scalar;
+
+    /// Type associated with elements of the space.
+    type E: Element<Space = Self>;
+
+    /// Create a new vector from the space.
+    fn create_element(&self) -> Self::E {
+        std::unimplemented!();
+    }
+
+    /// Norm of a vector.
+    fn norm(_x: &Self::E, _res: &mut <Self::F as cauchy::Scalar>::Real) -> Result {
+        Err(Error::NotImplemented)
+    }
+}
+
+pub trait DualSpace: LinearSpace {
+    type Space: LinearSpace<F = Self::F>;
+
+    fn dual_pairing(
+        &self,
+        x: &Self::E,
+        other: <Self::Space as LinearSpace>::E,
+        res: &mut Self::F,
+    ) -> Result;
```

Comment:
I think we just want `other` to be a reference, but maybe it can be `impl Element<Self::Space>`.

Reply:
Agree. This was an oversight.